src/docfx/lib/git/GitUtility.cs

```diff
@@ -35,6 +35,42 @@ public static (string url, string branch) GetGitRemoteInfo(string remoteHref)
             return (url, refSpec);
         }
 
+        /// <summary>
+        /// Execute git action with remote url with token and then reset the remote url back without token
+        /// </summary>
+        public static async Task WithToken(string cwd, string remoteUrl, string token, Func<string, Task> action)
+        {
+            if (string.IsNullOrEmpty(token))
+            {
+                await action(remoteUrl);
+                return;
+            }
+
+            // add token to remote href
+            var remoteUrlWithToken = AppendToken(remoteUrl, token);
+
+            await action(remoteUrlWithToken);
+
+            // reset url back without token
+            await SetRemoteGitUrl(cwd, remoteUrl);
```

Comment:
>await SetRemoteGitUrl(cwd, remoteUrl); [](start = 12, length = 38)

Is this only needed by `clone`? I see `fetch` doesn't use this. If so, can we put this into `GitUtility.Clone`?

Reply:
pull also need, it's just a wrapper for any action and i don't want to git clone itself

---
In reply to: [202904586](https://github.com/dotnet/docfx/pull/3085#discussion_r202904586) [](ancestors = 202904586)

Reply:
Got it. Personally, I prefer to hide the logic inside clone/pull, with an extra optional parameter passing the token.

Reply:
second @superyyrrzz

Reply:
With token logic will have an extra action to reset the url back, my purpose is to let the user choice, not a default logic hided inside clone/pull

---
In reply to: [202929623](https://github.com/dotnet/docfx/pull/3085#discussion_r202929623) [](ancestors = 202929623)

Reply:
For now the action == reset_url, so I think the clone/pull method itself already knows what action to take, thus to reduce user's effort to choose an action.